### DISCUSSION TITLE
HX 0.5: shortcut  = (format_selections) doesn't work / missing?

### DISCUSSION THREAD

**Discussion by Robert-M-Muench** [helix-editor/helix#1079] [Q&A]
I tried to format a selection with `=` but that doesn't do anything. Using command mode, I only see `format` but not `format_selections` 

Is there any trick how to use this feature?

**Comment by sudormrfbin** [comment]
Both `:format` and `format_selections` require an LSP active for the current file for it to work, I think it's not mentioned in the docs yet.

**Reply by Robert-M-Muench** [reply]
I added an LSP via a languages.toml section:

```
[[language]]
name = "javascript"
file-types = ["js", "mjs", "jsx"]
language-server = {command = "typescript-language-server --stdio"}
```

I assume that HX will start the language server. Is there a way to find out if this worked and an LSP connection is used?

**Reply by kirawi** [reply]
You can check the logs https://github.com/helix-editor/helix/wiki/FAQ#access-the-log-file

**Comment by archseer** [comment]
`=` is currently also disabled because it needed some work to do correctly with multiple selections.

https://github.com/helix-editor/helix/blob/fa0cb010e16c6a7d95b8849f073f472b0bd710be/helix-term/src/commands.rs#L4233-L4279

Note that a lot of LSPs also don't implement formatting with ranges (I don't think rust-analyzer does yet) so we should add a fallback that uses tree-sitter indentation (should be as easy as calculating an indent level for each line, and adding or deleting indentation based on that value).  https://github.com/helix-editor/helix/blob/fa0cb010e16c6a7d95b8849f073f472b0bd710be/helix-core/src/indent.rs#L483-L493

**Reply by Robert-M-Muench** [reply]
I think adding a simple indention formatter is good enough for 99% of all cases.

**Comment by MajorDallas** [comment]
If I may resurrect this one... I was about to make a new enhancement issue for this, but thought this might be better to try first.

I'm using Helix with some python code, the recommended LSP, and Black. `:format` works just fine as far as manipulating the text, but I have one gripe: it leaves the entire buffer selected so `;` puts the cursor at the beginning of the buffer, and if I didn't do a `save_selection` before running it I can't go back to my place with `C-o`.  I assume other LSP+formatters would exhibit the same behaviour, but haven't tested any.

Is this as-intended behaviour? Or would I be right to think of this as a bug?

**Reply by kirawi** [reply]
https://github.com/helix-editor/helix/issues/3859
